packages/vite/src/node/optimizer/optimizer.ts

```diff
@@ -516,6 +564,88 @@ async function createDepsOptimizer(
     }, timeout)
   }
 
+  async function onCrawlEnd() {
+    debug(colors.green(`✨ static imports crawl ended`))
+    if (firstRunCalled) {
+      return
+    }
+
+    currentlyProcessing = false
+
+    const crawlDeps = Object.keys(metadata.discovered)
+
+    // Await for the scan+optimize step running in the background
+    // It normally should be over by the time crawling of user code ended
+    await depsOptimizer.scanProcessing
+
+    if (!isBuild && postScanOptimizationResult) {
+      const result = await postScanOptimizationResult
+      postScanOptimizationResult = undefined
+
+      const scanDeps = Object.keys(result.metadata.optimized)
+
+      if (scanDeps.length === 0 && crawlDeps.length === 0) {
+        debug(
+          colors.green(
+            `✨ no dependencies found by the scanner or crawling static imports`
+          )
+        )
+        result.cancel()
+        firstRunCalled = true
+        return
+      }
+
+      const needsInteropMismatch = findInteropMismatches(
+        metadata.discovered,
+        result.metadata.optimized
+      )
+      const scannerMissedDeps = crawlDeps.some((dep) => !scanDeps.includes(dep))
+      const outdatedResult =
+        needsInteropMismatch.length > 0 || scannerMissedDeps
+
+      if (outdatedResult) {
+        // Drop this scan result, and perform a new optimization to avoid a full reload
+        result.cancel()
+
+        // Add deps found by the scanner to the discovered deps while crawling
+        for (const dep of scanDeps) {
+          if (!crawlDeps.includes(dep)) {
```

Comment:
Excuse me, why does the scanner dependency not exist in the crawling discovered dependency

Reply:
This line checks that when we crawl for dependencies, those that are in `scanDeps` (already optimized), shouldn't be added as a missing import since it's already optimized.

Reply:
There is one point that I don't quite understand.

Dependencies found during the scan phase will be stored in `metadata.discovered`, and then runOptimizeDeps will generate `result metadata.optimized(scanDeps)`. In the craw phase, `crawlDeps` should contain all the dependencies of the scan phase (`scanDeps`). Think this conditional statement will not hit, right?

Reply:
Sorry for the late reply. I think it's a bit out of my territory now since I haven't been following up on this part of the code. Maybe @patak-dev knows more about this. Also, are you suggesting of a perf improvement, or a bug?

Reply:
With the current code, it could still hit, but I think you have a point @Veath. We should move https://github.com/vitejs/vite/blob/881050d265a5c0d5c044ec1510339f7254082c21/packages/vite/src/node/optimizer/optimizer.ts#L572 
after the await of the scanner
https://github.com/vitejs/vite/blob/881050d265a5c0d5c044ec1510339f7254082c21/packages/vite/src/node/optimizer/optimizer.ts#L579
and then it should never hit. PR welcome if you would like to tackle that

Reply:
Can you explain under what circumstances will it hit?